scipy/stats/_distn_infrastructure.py

```diff
@@ -1261,22 +1260,34 @@ def moment(self, n, *args, **kwds):
                 mdict = {'moments': {1: 'm', 2: 'v', 3: 'vs', 4: 'vk'}[n]}
             else:
                 mdict = {}
-            mu, mu2, g1, g2 = self._stats(*goodargs, **mdict)
-        val = _moment_from_stats(n, mu, mu2, g1, g2, self._munp, goodargs)
+            mu, mu2, g1, g2 = self._stats(*shapes, **mdict)
+        val = np.empty(loc.shape)  # val needs to be indexed by loc
+        val[...] = _moment_from_stats(n, mu, mu2, g1, g2, self._munp, shapes)
 
         # Convert to transformed  X = L + S*Y
         # E[X^n] = E[(L+S*Y)^n] = L^n sum(comb(n, k)*(S/L)^k E[Y^k], k=0...n)
-        fac = _lazywhere(cond1_, (goodloc, goodscale),
-                         lambda loc, scale: scale / loc, self.badvalue)
-        for k in range(n):
-            valk = _moment_from_stats(k, mu, mu2, g1, g2, self._munp,
-                                      goodargs)
-            result[cond0] += comb(n, k, exact=True)*fac**k * valk
-        result[cond0] += fac**n * val
-        result[cond0] *= goodloc**n
-        res = goodscale**n * val
-        place(result, (cond0 & ~cond1), res[~cond1_])
-        place(result, ~cond0, self.badvalue)
+        result = zeros(i0.shape)
+        place(result, ~i0, self.badvalue)
+
+        if i1.any():
+            res1 = scale[loc == 0]**n * val[loc == 0]
+            place(result, i1, res1)
+
+        if i2.any():
+            if mu is not None:
+                mu, mu2, g1, g2 = argsreduce(loc != 0, mu, mu2, g1, g2)
+            args = argsreduce(loc != 0, *shapes, loc, scale, val)
+            shapes, loc, scale, val = args[:-2], args[-3], args[-2], args[-1]
```

Comment:
Hi, @mdhaber. I just noticed a small problem here. This should be `shapes, loc, scale, val = args[:-3], args[-3], args[-2], args[-1]` otherwise it passes extra arguments to the `_moment_from_stats` function and it throws a `TypeError`. I will push this change on my branch and add some tests for this!